File: app/Models/Subscription.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class Subscription extends Model
{
    use HasFactory;

    protected $fillable = [
        'user_id',
        'product_id',
        'plan_id',
        'account_id',
        'order_id',
        'start_date',
        'expiry_date',
        'status',
        'cancelled_at',
    ];

    protected $casts = [
        'start_date' => 'date',
        'expiry_date' => 'date',
        'cancelled_at' => 'datetime',
    ];

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function product(): BelongsTo
    {
        return $this->belongsTo(Product::class);
    }

    public function plan(): BelongsTo
    {
        return $this->belongsTo(Plan::class);
    }

    public function account(): BelongsTo
    {
        return $this->belongsTo(Account::class);
    }

    public function order(): BelongsTo
    {
        return $this->belongsTo(Order::class);
    }

    public function scopeActive($query)
    {
        return $query->where('status', 'active');
    }

    public function scopeExpiringSoon($query)
    {
        return $query->where('status', 'expiring_soon');
    }

    public function scopeExpired($query)
    {
        return $query->where('status', 'expired');
    }

    public function isActive(): bool
    {
        return $this->status === 'active';
    }

    public function isExpired(): bool
    {
        return now()->gt($this->expiry_date);
    }

    public function daysRemaining(): int
    {
        return (int) now()->diffInDays($this->expiry_date, false);
    }

    public function isExpiringSoon(int $thresholdDays = 3): bool
    {
        return $this->daysRemaining() <= $thresholdDays && $this->isActive();
    }
}

File: mhub/lib/models/subscription.dart
class Subscription {
  const Subscription({
    required this.id,
    required this.product,
    this.plan,
    required this.status,
    this.startDate,
    this.expiryDate,
    required this.expiresIn,
  });

  final int id;
  final String product;
  final String? plan;
  final String status;
  final String? startDate;
  final String? expiryDate;
  final String expiresIn;

  factory Subscription.fromJson(Map<String, dynamic> j) => Subscription(
        id: (j['id'] as num).toInt(),
        product: j['product'] as String? ?? '—',
        plan: j['plan'] as String?,
        status: j['status'] as String? ?? 'active',
        startDate: j['start_date'] as String?,
        expiryDate: j['expiry_date'] as String?,
        expiresIn: j['expires_in'] as String? ?? '',
      );
}
